### PR TITLE
feat(datastore): Return Get, GetMulti, Put and PutMulti errors with enhanced details

### PR DESCRIPTION
**Pull request by ionicc**
Closes: https://github.com/googleapis/google-cloud-go/issues/7060

## Changes

* Returns enhanced errors by adding more metadata in the errors
* As present in TODO: `checkMultiArgs()` logic has been folded in `Get()` and `Put()`.
* Unify errors by converting all `errors.New` to `fmt.Errorf`

### CODE REVIEW

File: datastore/datastore.go [hunk 1]
```diff
@@ -114,7 +113,7 @@ func NewClient(ctx context.Context, projectID string, opts ...option.ClientOptio
 	}
 
 	if projectID == "" {
-		return nil, errors.New("datastore: missing project/dataset id")
+		return nil, fmt.Errorf("datastore: missing project/dataset id")
```
Comment (telpirion) [inline datastore/datastore.go:116]: In places where no error is being wrapped, you don't need to replace `errors.New()`.

File: datastore/datastore.go [hunk 2]
```diff
@@ -134,19 +133,21 @@ func detectProjectID(ctx context.Context, opts ...option.ClientOption) (string,
 		return "", fmt.Errorf("fetching creds: %w", err)
 	}
 	if creds.ProjectID == "" {
-		return "", errors.New("datastore: see the docs on DetectProjectID")
+		return "", fmt.Errorf("datastore: see the docs on DetectProjectID")
```
Comment (telpirion) [inline datastore/datastore.go:136]: Same as previous.

File: datastore/datastore.go [hunk 3]
```diff
@@ -134,19 +133,21 @@ func detectProjectID(ctx context.Context, opts ...option.ClientOption) (string,
...
 	}
 	return creds.ProjectID, nil
 }
 
 var (
 	// ErrInvalidEntityType is returned when functions like Get or Next are
 	// passed a dst or src argument of invalid type.
-	ErrInvalidEntityType = errors.New("datastore: invalid entity type")
+	ErrInvalidEntityType = fmt.Errorf("datastore: invalid entity type")
```
Comment (telpirion) [inline datastore/datastore.go:144]: Same as previous.

File: datastore/datastore.go [hunk 4]
```diff
@@ -134,19 +133,21 @@ func detectProjectID(ctx context.Context, opts ...option.ClientOption) (string,
...
 	// ErrInvalidKey is returned when an invalid key is presented.
-	ErrInvalidKey = errors.New("datastore: invalid key")
+	ErrInvalidKey = fmt.Errorf("datastore: invalid key")
```
Comment (telpirion) [inline datastore/datastore.go:146]: Same as previous.

File: datastore/datastore.go [hunk 5]
```diff
@@ -134,19 +133,21 @@ func detectProjectID(ctx context.Context, opts ...option.ClientOption) (string,
...
 	// ErrNoSuchEntity is returned when no entity was found for a given key.
-	ErrNoSuchEntity = errors.New("datastore: no such entity")
+	ErrNoSuchEntity = fmt.Errorf("datastore: no such entity")
```
Comment (telpirion) [inline datastore/datastore.go:148]: Same as previous.

File: datastore/datastore.go [hunk 6]
```diff
@@ -134,19 +133,21 @@ func detectProjectID(ctx context.Context, opts ...option.ClientOption) (string,
...
+	// ErrDifferentKeyAndDstLength is returned when the length of dst and key are different.
+	ErrDifferentKeyAndDstLength = fmt.Errorf("datastore: keys and dst slices have different length")
```
Comment (telpirion) [inline datastore/datastore.go:150]: Same as previous.

**Review by telpirion — CHANGES_REQUESTED** [review]
Please revert the unnecessary replacements of `errors.New()` and I'll take another look. Thank you!

File: datastore/datastore.go (hunk 1 above)
Comment (ionicc) [reply datastore/datastore.go]: @telpirion Thank you for the review. I replaced it for consistency, But I've reverted it now. Please take a look!

**Comment by ionicc** [comment]
@telpirion The comments have been resolved, Could you check this PR again?

**Review by telpirion — APPROVED** [review]

**Pull request merged**